src/NumericalAlgorithms/SphericalHarmonics/IO/ReadSurfaceYlm.cpp

```diff
@@ -0,0 +1,194 @@
+  return n == abs(n) and n == floor(n);
+}
+
+template <typename Frame>
+Strahlkorper<Frame> read_surface_ylm_row(const Matrix& ylm_data,
+                                         const size_t row_number) {
+  const size_t l_max_column_number = 4;
+  const std::array<double, 3> expansion_center{{ylm_data(row_number, 1),
+                                                ylm_data(row_number, 2),
+                                                ylm_data(row_number, 3)}};
+  const double l_max_from_file = ylm_data(row_number, l_max_column_number);
+  if (not is_nonnegative_int(l_max_from_file)) {
+    ERROR("Row " << row_number << " of the Ylm data has an invalid Lmax value ("
+                 << l_max_from_file << ") in column " << l_max_column_number
+                 << ". The value of Lmax should be a nonnegative integer.");
+  }
+
+  const size_t l_max = ylm_data(row_number, l_max_column_number);
+  // l_max == m_max
+  const size_t spectral_size = Spherepack::spectral_size(l_max, l_max);
+  // We only write and store half of the coefficients. This is the minimum
+  // number of coefficients needed to describe the strahlkorper given the l_max,
+  // i.e. columns of 0.0 for higher coefficients are okay
+  const size_t min_expected_num_coefficients = spectral_size / 2;
+  const size_t num_non_coef_headers = 5;
+  const size_t min_expected_num_columns =
+      min_expected_num_coefficients + num_non_coef_headers;
+  const size_t actual_num_columns = ylm_data.columns();
+
+  const std::string frame{get_output(Frame{})};
+  const std::string expected_format{
+      MakeString{}
+      << "The expected format of the data is \'Time, " << frame
+      << "ExpansionCenter_x, " << frame << "ExpansionCenter_y, " << frame
+      << "Expansion_Center_z, Lmax, coef(0,0), coef(1,-1), coef(1,0), "
+         "coef(1,1), coef(2,-2), coef(2,-1), coef(2,0), coef(2,1), coef(2,2), "
+         "..., coef(Lmax,Lmax), [0.0...]\', where the number of coefficients "
+         "is equal to (Lmax + 1)^2 and the coefficient columns may be padded "
+         "with columns of 0.0 for any higher order coefficients for l > Lmax."};
+  if (actual_num_columns < min_expected_num_columns) {
+    ERROR("Row "
+          << row_number
+          << " of the Ylm data does not have the expected format. For Lmax = "
+          << l_max << ", expected at least " << min_expected_num_columns
+          << " columns.\n\n"
+          << expected_format);
+  }
+
+  ModalVector spectral_coefficients(spectral_size, 0.0);
+  size_t coef_column_number = num_non_coef_headers;
+  SpherepackIterator iter(l_max, l_max);
+  // read in expected coefficients for the given l_max
+  for (size_t l = 0; l <= l_max; l++) {
+    for (int m = -static_cast<int>(l); m <= static_cast<int>(l); m++) {
+      iter.set(l, m);
+
+      const double coefficient = ylm_data(row_number, coef_column_number);
+      spectral_coefficients[iter()] = coefficient;
+
+      coef_column_number++;
```

Comment:
This relies on the order of the columns. Please check the legend to make sure that the coefficient you are reading corresponds to the `(l,m)` that you are expecting here. (It would be quite hard to debug if these numbers were read in the wrong order.)

Reply:
Maybe do this in the calling function `read_surface_ylm` so you don't have to check for every row of data?